File: validation/roundtrip/backend.js
#!/usr/bin/env node
'use strict';
// Identifiable test backend for the behavioral round-trip lab.
// Every response reports which backend served it, so the prober can measure
// load-balancing distribution and persistence through any LB in front.
//
//   node backend.js <id> [port=8080]
//
// Endpoints:
//   GET /            -> JSON {backend, port, path, ts}; also echoed in the
//                       X-Backend header (works for HEAD too)
//   GET /health      -> 200 "OK <id>" while up, 503 while administratively down
//   POST /admin/down -> health starts returning 503 (LB should eject this backend)
//   POST /admin/up   -> health returns 200 again
//
// /admin/* is meant to be called DIRECTLY (not through the LB) by probe.js to
// exercise the LB's health-check reaction.
const http = require('http');
const id = process.argv[2] || 'backend';
const port = parseInt(process.argv[3] || '8080', 10);
let healthy = true;

http.createServer((req, res) => {
  const respond = (code, body, type) => {
    res.writeHead(code, { 'Content-Type': type || 'application/json', 'X-Backend': id });
    res.end(body);
  };
  if (req.url === '/health') {
    return respond(healthy ? 200 : 503, (healthy ? 'OK ' : 'DOWN ') + id, 'text/plain');
  }
  if (req.method === 'POST' && req.url === '/admin/down') { healthy = false; return respond(200, 'health=down ' + id, 'text/plain'); }
  if (req.method === 'POST' && req.url === '/admin/up') { healthy = true; return respond(200, 'health=up ' + id, 'text/plain'); }
  respond(200, JSON.stringify({ backend: id, port, path: req.url, ts: Date.now() }));
}).listen(port, () => console.log('backend ' + id + ' listening on :' + port));

File: validation/roundtrip/backend.py
#!/usr/bin/env python3
"""Identifiable test backend — Python twin of backend.js for hosts without
Node (e.g. lab VMs with no internet for installs). Same contract:
GET / -> JSON + X-Backend header; GET /health -> 200/503 toggled by
POST /admin/down|up.   usage: python3 backend.py <id> <port> [bind-ip]"""
import sys, json, time
from http.server import BaseHTTPRequestHandler, HTTPServer

ID = sys.argv[1] if len(sys.argv) > 1 else 'backend'
PORT = int(sys.argv[2]) if len(sys.argv) > 2 else 8080
BIND = sys.argv[3] if len(sys.argv) > 3 else '0.0.0.0'
healthy = True

class H(BaseHTTPRequestHandler):
    def log_message(self, *a):
        pass
    def _send(self, code, body, ctype='application/json'):
        data = body.encode()
        self.send_response(code)
        self.send_header('Content-Type', ctype)
        self.send_header('Content-Length', str(len(data)))
        self.send_header('X-Backend', ID)
        self.end_headers()
        self.wfile.write(data)
    def do_GET(self):
        if self.path == '/health':
            return self._send(200 if healthy else 503,
                              ('OK ' if healthy else 'DOWN ') + ID, 'text/plain')
        self._send(200, json.dumps({'backend': ID, 'port': PORT,
                                    'path': self.path, 'ts': int(time.time() * 1000),
                                    'client': self.client_address[0],
                                    'xff': self.headers.get('X-Forwarded-For')}))
    def do_POST(self):
        global healthy
        if self.path == '/admin/down':
            healthy = False
            return self._send(200, 'health=down ' + ID, 'text/plain')
        if self.path == '/admin/up':
            healthy = True
            return self._send(200, 'health=up ' + ID, 'text/plain')
        self._send(404, 'unknown', 'text/plain')

if __name__ == '__main__':
    print('backend %s listening on %s:%d' % (ID, BIND, PORT), flush=True)
    HTTPServer((BIND, PORT), H).serve_forever()
